data-serving/data-service/src/controllers/case.ts

```diff
@@ -136,6 +136,114 @@ export const batchValidate = async (
     }
 };
 
+/**
+ * Find IDs of existing cases that have {caseReference.sourceId,
+ * caseReference.sourceEntryId} combinations matching any cases in the provided
+ * request.
+ *
+ * This is used in batchUpsert. Background:
+ *
+ *   While MongoDB does return IDs of created documents, it doesn't do so
+ *   for modified documents (e.g. cases updated via upsert calls). In
+ *   order to (necessarily) provide that information, we'll query existing
+ *   cases, filtering on provided case reference data, in order to provide
+ *   an accurate list of updated case IDs.
+ */
+const findCaseIdsWithCaseReferenceData = async (
+    req: Request,
+): Promise<string[]> => {
+    const providedCaseReferenceData = req.body.cases
+        .filter(
+            // Case data should be validated prior to this point.
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            (c: any) =>
+                c.caseReference?.sourceId && c.caseReference?.sourceEntryId,
+        )
+        .map((c: any) => {
+            return {
+                'caseReference.sourceId': c.caseReference.sourceId,
+                'caseReference.sourceEntryId': c.caseReference.sourceEntryId,
+            };
+        });
+    const result =
+        providedCaseReferenceData.length > 0
+            ? await Case.find()
+                  .select('_id')
+                  .or(providedCaseReferenceData)
+                  .lean()
+                  .exec()
+            : [];
+    return result.map((res) => String(res['_id']));
+};
+
+/**
+ * Batch upserts cases.
+ *
+ * Handles HTTP POST /api/cases/batchUpsert.
+ *
+ * Note that this method is _not_ atomic, and that validation _should_ be
+ * performed prior to invocation. Upserted cases are not validated, and while
+ * any validation issues for created cases will cause the API to return 422,
+ * all provided cases without validation errors will be written.
+ *
+ * TODO: Wrap batchValidate in this method.
+ */
+export const batchUpsert = async (
+    req: Request,
+    res: Response,
+): Promise<void> => {
+    try {
+        const toBeUpsertedCaseIds = await findCaseIdsWithCaseReferenceData(req);
```

Comment:
if this still isn't as fast as you want, in theory you could have this fn add the ids to the requests, so that below you don't have to re-filter? (assuming that's even an option in mongodb's bulkwrite fn)

Reply:
Meaning to perform the updateOne()'s specifying the _id, rather than the caseRef data pair?

I think that's a solid idea. I'm a little worried about this step being generally slow, but plan to test and make these kinds of tweaks in the subsequent change.